lib/Newsletter/NewsletterSaveController.php

```diff
@@ -119,61 +111,114 @@ public function save(array $data = []): array {
       // if newsletter was previously scheduled and is now unscheduled, set its status to DRAFT and delete associated queue record
-      if ($newsletter->status === Newsletter::STATUS_SCHEDULED && isset($options['isScheduled']) && empty($options['isScheduled'])) {
+      if ($newsletterModel->status === Newsletter::STATUS_SCHEDULED && isset($options['isScheduled']) && empty($options['isScheduled'])) {
         $queue->delete();
-        $newsletter->status = Newsletter::STATUS_DRAFT;
-        $newsletter->save();
+        $newsletterModel->status = Newsletter::STATUS_DRAFT;
+        $newsletterModel->save();
       } else {
         $queue->newsletterRenderedBody = null;
         $queue->newsletterRenderedSubject = null;
         $newsletterQueueTask = new NewsletterQueueTask();
-        $newsletterQueueTask->preProcessNewsletter($newsletter, $queue);
+        $newsletterQueueTask->preProcessNewsletter($newsletterModel, $queue);
       }
     }
 
-    $this->wp->doAction('mailpoet_api_newsletters_save_after', $newsletter);
-    $this->authorizedEmailsController->onNewsletterUpdate($newsletter, $oldNewsletter);
+    $this->wp->doAction('mailpoet_api_newsletters_save_after', $newsletterModel);
+    $this->authorizedEmailsController->onNewsletterUpdate($newsletterModel, $oldNewsletterModel);
+
+    $previewUrl = NewsletterUrl::getViewInBrowserUrl($newsletterModel);
+    return [$newsletterModel->asArray(), ['preview_url' => $previewUrl]];
+  }
+
+  private function getNewsletter(array $data): NewsletterEntity {
+    if (!isset($data['id'])) {
+      throw new UnexpectedValueException();
+    }
+
+    $newsletter = $this->newslettersRepository->findOneById((int)$data['id']);
+    if (!$newsletter) {
+      throw new NotFoundException();
+    }
+    return $newsletter;
+  }
+
+  private function updateNewsletter(NewsletterEntity $newsletter, array $data) {
+    if (array_key_exists('type', $data)) {
+      $newsletter->setType($data['type']);
+    }
+
+    if (array_key_exists('subject', $data)) {
+      $newsletter->setSubject($data['subject']);
+    }
+
+    if (array_key_exists('preheader', $data)) {
+      $newsletter->setPreheader($data['preheader']);
+    }
+
+    if (array_key_exists('body', $data)) {
+      $newsletter->setBody(json_decode($data['body'], true));
+    }
+
+    if (array_key_exists('ga_campaign', $data)) {
+      $newsletter->setGaCampaign($data['ga_campaign']);
+    }
+
+    if (array_key_exists('sender_name', $data)) {
+      $newsletter->setSenderName($data['sender_name'] ?? '');
+    }
+
+    if (array_key_exists('sender_address', $data)) {
+      $newsletter->setSenderAddress($data['sender_address'] ?? '');
+    }
+
+    if (array_key_exists('reply_to_name', $data)) {
+      $newsletter->setReplyToName($data['reply_to_name'] ?? '');
+    }
+
+    if (array_key_exists('reply_to_address', $data)) {
+      $newsletter->setReplyToAddress($data['reply_to_address'] ?? '');
+    }
```

Comment:
In the old [model's createUpdate is a functionality](https://github.com/mailpoet/mailpoet/blob/master/lib/Models/Newsletter.php#L990-L1019) that sets these properties based on values from settings instead of an empty string as a fallback. I'm not sure if removing this may cause some issues. Maybe we should keep it. What do you think?

Reply:
I thought so too but it was misbehaving... and the reason is the code you point to is only for `onCreate`. Note that the `save` endpoint is update-only now (we never use it for create, there's a different endpoint for that).